service/history/queues/executable.go

```diff
@@ -234,6 +208,32 @@ func (e *executableImpl) Execute() (retErr error) {
 		// Not doing it here as for certain errors latency for the attempt should not be counted
 	}()
 
+	e.Lock()
+	if e.state != ctasks.TaskStatePending {
+		e.Unlock()
+		return nil
+	}
```

Comment:
can we move this part above the defer? don't think we need metrics for tasks that was cancelled. The lock is just for getting state & priority, and in one lock we can get and remember both values.